### ISSUE TITLE
Opening a side module larger than 4KB fails on Chrome

### ISSUE THREAD

**Issue by cuinjune** [emscripten-core/emscripten#11753]
Hi, I'm trying to dynamically load a side module from the main module using `dlopen()`.
The side module loads fine as long as its size is smaller than 4KB but I need to load large-size side modules.
Here's a simple code you can test this:

**side.c:**
```
#define SIZE 4000
char dummy[SIZE] = {};

int side(int a)
{
    return SIZE;
}
```
**main.c:**
```
#include <stdio.h>
#include <dlfcn.h>

int main() 
{
    void *handle;
    typedef int (*func_t)(int);

    handle = dlopen("side.wasm", RTLD_NOW);

    if (!handle) {
        printf("failed to open the library\n");
        return 0;
    }
    func_t func = (func_t)dlsym(handle, "side");

    if (!func) {
        printf("failed to find the method\n");
        dlclose(handle);
        return 0;
    }
    printf("side module size: %d byte\n", func(1));
}
```
**index.html:**
```
<!DOCTYPE html>
<html>

<head>
</head>

<body>
  <script async src="main.js"></script>
</body>

</html>
```

**commands:**
```
emcc side.c -s SIDE_MODULE=1 -o side.wasm
emcc main.c -s MAIN_MODULE=1 -o main.html --preload-file side.wasm
python3 -m http.server 8080
```
And this is the result I get in the Chrome browser:

> Error in loading dynamic library side.wasm: RangeError: WebAssembly.Compile is disallowed on the main thread, if the buffer size is larger than 4KB. Use WebAssembly.compile, or compile on a worker thread.

Can someone please guild me on how to dynamically load a side module larger than 4KB?
Thank you in advance!

**Comment by gerald-dotcom** [comment]
It's a Chrome limitation, you can use `loadDynamicLibrary(url, {loadAsync: true, global: true, nodelete: true})` but there is one thing that I'm missing to make it work. Maybe someone can help you out.

**Comment by kripken** [comment]
It looks like the preloading logic looks for the normal shared library suffix on POSIX, `.so`,

https://github.com/emscripten-core/emscripten/blob/df56ba11fe8dbae3949446239affe69b2efa7772/src/library_browser.js#L241

We should document that if it isn't already.

**Comment by cuinjune** [comment]
@sbc100 Could you please help me?

**Comment by sbc100** [comment]
Can you try using the `.so` extension as @kripken suggested?  It looks like there is some special handling to preload side modules but it looks for the `.so` extension.

BTW in this mode you are going to delay the startup of you program until both the side module and main module are instantiated, so you are not really getting any benefit at runtime from using the a side module.   Would it make more sense to simply compile the side module into the main program?

**Comment by cuinjune** [comment]
@sbc100 How do I build a side module as the `so` extension?
Can I just replace `wasm` with `so` like the following?

`emcc side.c -s SIDE_MODULE=1 -o side.so`

I need the side module to exist separately from the main module, and the side module will be instantiated in the middle of the program not necessarily at the startup.

**Comment by sbc100** [comment]
Yes `-o side.so` should work.  Better still. `-o libside.so` to match the UNIX convension.

If you preload the file with `--preload-file` then the side module will actually be instantiated at startup.. even though logically you don't access it until later.   It should be be observable to you program other than the that the startup will be delayed until the side module has been downloaded and compiled.   Does the work for you?

**Comment by cuinjune** [comment]
I'm trying right now, I could build the side module with `emcc side.c -s SIDE_MODULE=1 -c -o side.so`

**Comment by cuinjune** [comment]
I used these commands to build:

```
emcc side.c -s SIDE_MODULE=1 -c -o side.so
emcc main.c -s MAIN_MODULE=1 -o main.html --preload-file side.so
python3 -m http.server 8080
```

Now I got these errors when I run it in Chrome:

> Assertion failed: need the dylink section to be first
> 
> Error in loading dynamic library side.so: RuntimeError: abort(Assertion failed: need the dylink section to be first) at Error
>     at jsStackTrace (main.js:2568)
>     at stackTrace (main.js:2586)
>     at abort (main.js:2288)
>     at assert (main.js:1329)
>     at loadWebAssemblyModule (main.js:756)
>     at createLibModule (main.js:650)
>     at getLibModule (main.js:669)
>     at loadDynamicLibrary (main.js:708)
>     at _dlopen (main.js:7686)
>     at __original_main (:8080/<anonymous>:wasm-function[153]:0x6c7e7)

**Comment by cuinjune** [comment]
If loading the large-size side module with `dlopen()` is not possible, is there any workaround to this? (e.g. loading the `.wasm` file from `index.html`?)

**Comment by sbc100** [comment]
Did you remember to change the name of side module in main.c?

Its should work with reloading such that side.so is loaded at start and you should not see a runtime call to .   `loadWebAssemblyModule` on dlopen becasue `side.so` is already included in `Module['preloadedWasm']`.   See https://github.com/emscripten-core/emscripten/blob/8bd9c9379d5697c06b72d549dfc28f47ff6fcb6c/src/library_browser.js#L252

If that doesn't work you can try using the asynrous JS function  loadDynamicLibrary as suggested earlier in this thread. 
 However the dlopen+preload approach should work if you want it to.

**Comment by cuinjune** [comment]
@sbc100 Yes I did of course.
Here's what everything looks like:

**side.c:**
```
#define SIZE 3000
char dummy[SIZE] = {};

int side(int a)
{
    return SIZE;
}
```
**main.c:**
```
#include <stdio.h>
#include <dlfcn.h>

int main()
{
    void *handle;
    typedef int (*func_t)(int);

    handle = dlopen("side.so", RTLD_NOW);

    if (!handle) {
        printf("failed to open the library\n");
        return 0;
    }
    func_t func = (func_t)dlsym(handle, "side");

    if (!func) {
        printf("failed to find the method\n");
        dlclose(handle);
        return 0;
    }
    printf("side module size: %d byte\n", func(1));
}
```
**index.html:**
```
<!DOCTYPE html>
<html>

<head>
</head>

<body>
  <script async src="main.js"></script>
</body>

</html>
```
**commands:**
```
emcc side.c -s SIDE_MODULE=1 -c -o side.so
emcc main.c -s MAIN_MODULE=1 -o main.html --preload-file side.so
python3 -m http.server 8080
```
**Result:**

> Assertion failed: need the dylink section to be first
> 
> Error in loading dynamic library side.so: RuntimeError: abort(Assertion failed: need the dylink section to be first) at Error
> at jsStackTrace (main.js:2568)
> at stackTrace (main.js:2586)
> at abort (main.js:2288)
> at assert (main.js:1329)
> at loadWebAssemblyModule (main.js:756)
> at createLibModule (main.js:650)
> at getLibModule (main.js:669)
> at loadDynamicLibrary (main.js:708)
> at _dlopen (main.js:7686)
> at __original_main (:8080/:wasm-function[153]:0x6c7e7)

**Comment by cuinjune** [comment]
I tried to use the `loadDynamicLibrary`:

```
<!DOCTYPE html>
<html>

<head>
</head>

<body>
  <script>
	loadDynamicLibrary("side.wasm", {loadAsync: true, global: true, nodelete: true})
  </script>
  <script async src="main.js"></script>
</body>

</html>
```

But I get this error:

> Uncaught ReferenceError: loadDynamicLibrary is not defined

**Comment by gerald-dotcom** [comment]
> I tried to use the `loadDynamicLibrary`:
> 
> ```
> <!DOCTYPE html>
> <html>
> 
> <head>
> </head>
> 
> <body>
>   <script>
> 	loadDynamicLibrary("side.wasm", {loadAsync: true, global: true, nodelete: true})
>   </script>
>   <script async src="main.js"></script>
> </body>
> 
> </html>
> ```
> 
> But I get this error:
> 
> > Uncaught ReferenceError: loadDynamicLibrary is not defined

Use EM_JS and rename .wasm to .so

**Comment by cuinjune** [comment]
@gerald-dotcom I still get the same error:

> Uncaught (in promise) RuntimeError: abort(Assertion failed: need the dylink section to be first) at Error
>     at jsStackTrace (http://localhost:8080/main.js:2568:17)
>     at stackTrace (http://localhost:8080/main.js:2586:16)
>     at abort (http://localhost:8080/main.js:2288:44)
>     at assert (http://localhost:8080/main.js:1329:5)
>     at loadWebAssemblyModule (http://localhost:8080/main.js:756:3)
>     at createLibModule (http://localhost:8080/main.js:650:12)
>     at http://localhost:8080/main.js:665:16
>     at abort (http://localhost:8080/main.js:2294:11)
>     at assert (http://localhost:8080/main.js:1329:5)
>     at loadWebAssemblyModule (http://localhost:8080/main.js:756:3)
>     at createLibModule (http://localhost:8080/main.js:650:12)
>     at http://localhost:8080/main.js:665:16

Here's my new code:

**main.c:**
```
#include <stdio.h>
#include <dlfcn.h>
#include <emscripten.h>

EM_JS(void, loadLibrary, (), {
      loadDynamicLibrary("side.so", {loadAsync: true, global: true, nodelete: true});
});

int main()
{
    loadLibrary();
}
```
**commands:**
```
emcc side.c -s SIDE_MODULE=1 -c -o side.so
emcc main.c -s MAIN_MODULE=1 -o main.html --preload-file side.so
python3 -m http.server 8080
```

**Comment by sbc100** [comment]
Strange... somehow the side module is missing the dylink section which emscripten adds to the beginning of it.

**Comment by cuinjune** [comment]
I also posted my question on Stackoverflow https://stackoverflow.com/q/63150966/5224286 with a bounty.

**Comment by sbc100** [comment]
Ah!  The problem is that you are passing `-c` in `emcc side.c -s SIDE_MODULE=1 -c -o libside.so` which means "compile to an object file".    This means the resulting `side.so` is not a shared libray but just an object file.

If you remove `-c` it should work.

Sadly it looks like you also have to change the output name back to `.wasm` to persuade emcc to produce a side module.  We should fix that part.

**Comment by cuinjune** [comment]
@sbc100 But if I remove `-c` then it doesn't compile.
**command:**
`emcc side.c -s SIDE_MODULE=1 -o side.so`

I get the following error:

> emcc: error: SIDE_MODULE must only be used when compiling to an executable shared library, and not when emitting an object file.  That is, you should be emitting a .wasm file (for wasm) or a .js file (for asm.js). Note that when compiling to a typical native suffix for a shared library (.so, .dylib, .dll; which many build systems do) then Emscripten emits an object file, which you should then compile to .wasm or .js with SIDE_MODULE.

**Comment by sbc100** [comment]
Sadly it looks like you also have to change the output name back to .wasm to persuade emcc to produce a side module.  You will then need to rename the file .so when to preload it.   We should fix that part.

**Comment by sbc100** [comment]
Apologies for that very hard to understand error message... the situation with linking `.so` and `.dylib` extensions is complicated due to the history emscripten and keeping compataiblity with build systems that want to use those extensions but *don't* want side modules.

**Comment by cuinjune** [comment]
So is it not possible to load a side module larger than 4KB at the moment?

**Comment by sbc100** [comment]
No, its possible, you just need to do the instantiation asynchronously because the browser doesn't support the synchronous loading of larger modules.   This limitation is built into the browser and not related to emscripten.

Both loadDynamicLibrary (with `loadAsync: true`) and dlopen (in combination with the emscripten feature that pre-loads `.so` files) should allow for this asynchronous loading.

**Comment by cuinjune** [comment]
Thank you. so how can I generate the `.so` file?
Should I `emcc side.c -s SIDE_MODULE=1 -o side.wasm` and then rename the `side.wasm` to `side.so`?

**Comment by sbc100** [comment]
Yes, if you want to use dlopen in combination with the preload plugin system that seems necessary today.

I will work two changes:

1. Allow linking side modules directly to `.so`.
2. Allow `.wasm` files to be preloaded by the preload plugin.

But you should be able to get it working even without either of those changes.

**Comment by cuinjune** [comment]
Thanks, I tried it as you said. Here's my code:

**main.cpp**
```
#include <stdio.h>
#include <dlfcn.h>
#include <emscripten.h>
#include <emscripten/bind.h>

using namespace emscripten;

int side(int a);

void hello() {
    printf("side module size: %d byte\n", side(1));
}

EMSCRIPTEN_BINDINGS(my_module) {
    function("hello", &hello);
}

EM_JS(void, loadLibrary, (), {
      loadDynamicLibrary("side.so", {loadAsync: true, global: true, nodelete: true});
});

int main()
{
    loadLibrary();
}
```
**index.html**
```
<!DOCTYPE html>
<html>

<head>
</head>

<body>
  <button id="buttonPressed">call side() function</button>
  <script>
    var Module
      = {
      preRun: []
      , postRun: []
      , print: function (e) {
        1 < arguments.length && (e = Array.prototype.slice.call(arguments).join(" "));
        console.log(e);
      }
      , printErr: function (e) {
        1 < arguments.length && (e = Array.prototype.slice.call(arguments).join(" "));
        console.error(e)
      }
    };

    function buttonPressed() {
      Module.hello();
    }

    window.onload = async () => {
      document.getElementById("buttonPressed").addEventListener("click", buttonPressed, false);
    };
  </script>
  <script async src="main.js"></script>
</body>

</html>
```
**commands:**
```
emcc side.c -s SIDE_MODULE=1 -o side.wasm
mv -f side.wasm side.so
emcc --bind main.cpp -s MAIN_MODULE=1 -o main.html --preload-file side.so
python3 -m http.server 8080
```
After the page is loaded, if I press the `call side() function` button, I get the following error:

> main.js:2294 Uncaught RuntimeError: abort(external function '_Z4sidei' is missing. perhaps a side module was not linked in? if this function was expected to arrive from a system library, try to build the MAIN_MODULE with EMCC_FORCE_STDLIBS=1 in the environment) at Error

What did I do wrong?

**Comment by sbc100** [comment]
That looks like a name mangling issue.   If you compile your side module as C rather than C++ then you need to also add `extern "C"` to the declaration in your C++ code (main.cpp):

```
extern "C" {
  int side(int a);
}
```

**Comment by sbc100** [comment]
While working on some fixed in this area I noticed that the preload+dlopen approach has some other issues so its good that proceed with the loadDynamicLibrary approach for now until we can address them.

**Comment by cuinjune** [comment]
It worked! Thank you so much!

May I ask how to implement a callback function to detect if the `.so` file is successfully loaded?

**Comment by sbc100** [comment]
Looks like you can wait on the returned promise:

```
// - if flags.loadAsync=true, the loading is performed asynchronously and        
//   loadDynamicLibrary returns corresponding promise.     
```

**Comment by cuinjune** [comment]
Thank you so much!
Is it not possible to dynamically call a function from side module without declaring  `int side(int a)` in main module? This is the reason why I was using `dlopen()`.

**Comment by sbc100** [comment]
Try using `dlsym` with `RTLD_DEFAULT` is the first argument.  This will search global namespace of all symbols and should contains your `side` symbol after the code has been loaded.

**Comment by cuinjune** [comment]
Thank you very much! I will try that.

Can I ask you one more question?
My app should work synchronously, so it should pause all other processes until the `side.so` is loaded then the `side()` function is called. Would it work synchronously if I use `async/await`?

Here's my example code that works:

```
#include <stdio.h>
#include <emscripten.h>
#include <emscripten/bind.h>

using namespace emscripten;

extern "C" {
  int side(int a);
}

void hello() {
    printf("side module size: %d byte\n", side(1));
}

EMSCRIPTEN_BINDINGS(my_module) {
    function("hello", &hello);
}

EM_JS(void, loadLibrary, (), {
      async function doLoadLibrary() {
        try {
          await loadDynamicLibrary("side.so", {loadAsync: true, global: true, nodelete: true});
          Module.hello();
        }
        catch(error) {
          console.log(error);
        }
      }
      doLoadLibrary();
});

int main()
{
    loadLibrary();
    // do something after side(1) is called. (Can I expect this to always happen after the side(1) is called?)
}
```

**Comment by gerald-dotcom** [comment]
> Thank you very much! I will try that.
> 
> Can I ask you one more question?
> My app should work synchronously, so it should pause all other processes until the `side.so` is loaded then the `side()` function is called. Would it work synchronously if I use `async/await`?
> 
> Here's my example code that works:
> 
> ```
> #include <stdio.h>
> #include <emscripten.h>
> #include <emscripten/bind.h>
> 
> using namespace emscripten;
> 
> extern "C" {
>   int side(int a);
> }
> 
> void hello() {
>     printf("side module size: %d byte\n", side(1));
> }
> 
> EMSCRIPTEN_BINDINGS(my_module) {
>     function("hello", &hello);
> }
> 
> EM_JS(void, loadLibrary, (), {
>       async function doLoadLibrary() {
>         try {
>           await loadDynamicLibrary("side.so", {loadAsync: true, global: true, nodelete: true});
>           Module.hello();
>         }
>         catch(error) {
>           console.log(error);
>         }
>       }
>       doLoadLibrary();
> });
> 
> int main()
> {
>     loadLibrary();
>     // do something after side(1) is called. (Can I expect this to always happen after the side(1) is called?)
> }
> ```
> 
> **ADDED:**
> I just checked that if I print something after `loadLibrary()` then it gets printed to the console first before `side(1)` is called so I guess my approach doesn't work.
> 
> What I want is, for example if I load `a.so` and then `b.so`, it should be guaranteed that `a()` is called before `b()`.

EM_JS creates function behind the scenes. To use async await syntax, you have to use Asyncify but last time I checked, I can't compile https://github.com/emscripten-core/emscripten/issues/11717 so I went with callbacks  that use Asyncify  but with handle sleep. ..

**Comment by cuinjune** [comment]
@gerald-dotcom Thank you so much! Using Asyncify worked for me!
Here's my working code:

**main.c**
```
#include <stdio.h>
#include <emscripten.h>

int side(int a);

EM_JS(void, loadLibrary, (), {
      Asyncify.handleAsync(async () => {
        try {
          await loadDynamicLibrary("side.so", {loadAsync: true, global: true, nodelete: true});
        }
        catch(error) {
          console.log(error);
        }
      });
});

int main()
{
    loadLibrary();
    printf("side module size: %d byte\n", side(1));
}
```
**commands:**
`emcc main.c -s MAIN_MODULE=1 -s ASYNCIFY -s 'ASYNCIFY_IMPORTS=["loadLibrary"]' -o main.html --preload-file side.so`

**Comment by cuinjune** [comment]
@sbc100 I tried using `dysym` with `RTLD_DEFAULT` but it doesn't work.

**main.c:**
```
#include <stdio.h>
#include <dlfcn.h>
#include <emscripten.h>

EM_JS(void, loadLibrary, (), {
      Asyncify.handleAsync(async () => {
        try {
          await loadDynamicLibrary("side.so", {loadAsync: true, global: true, nodelete: true});
        }
        catch(error) {
          console.log(error);
        }
      });
});

int main()
{
    loadLibrary();
    typedef int (*func_t)(int);
    func_t func = (func_t)dlsym(RTLD_DEFAULT, "side");
    if (!func) {
        printf("failed to find the method\n");
        return 0;
    }
    printf("side module size: %d byte\n", func(1));
}
```
**result:**

> failed to find the method

I think I did everything correctly. Any idea why it doesn't work?

**Comment by sbc100** [comment]
Ah, I guess emscripten does not support dlsym.

You could fix this by adding support for dlsym, or if it works for you you could call the function via `EM_JS` since `Module['side']` should contain this function after you load the code.

**Comment by cuinjune** [comment]
@sbc100 Thank you so much!  I could finally load a side module larger than 4KB and call its function synchronously.

Here's my full working code:

**side:**
```
#define SIZE 5000
char dummy[SIZE] = {};

int side()
{
    return SIZE;
}
```
**main.c:**
```
#include <stdio.h>
#include <dlfcn.h>
#include <emscripten.h>

EM_JS(void, loadLibrary, (const char *name), {
      Asyncify.handleAsync(async () => {
        try {
          var str = UTF8ToString(name);
          await loadDynamicLibrary(str + '.wasm', {loadAsync: true, global: true, nodelete: true});
            console.log('side module size: ' + Module['_' + str]());
        }
        catch(error) {
          console.log(error);
        }
      });
});

int main()
{
    printf("before\n");
    loadLibrary("side");
    printf("after\n");
}
```
**index.html:**
```
<!DOCTYPE html>
<html>

<head>
</head>

<body> 
  <script async src="main.js"></script>
</body>

</html>
```
**commands:**
```
emcc side.c -s SIDE_MODULE=1 -o side.wasm
emcc main.c -s MAIN_MODULE=1 -s ASYNCIFY -s 'ASYNCIFY_IMPORTS=["loadLibrary"]' -o main.html --preload-file side.wasm
python3 -m http.server 8080
```
**result:**

> before
> side module size: 5000
> after

**Comment by cuinjune** [comment]
@sbc100 I just found out `loadDynamicLibrary()` only works if the side module is preloaded. 
If I dynamically upload a side module using `Module["FS_createDataFile"]()` and try to load it, I get the following error:
> failed to load binary file at '/side.wasm'

However, loading the dynamically-uploaded side module works fine if I use `dlopen()` instead of `loadDynamicLibrary()` as long as the file size is smaller than 4KB.

Can `loadDynamicLibrary()` only load preloaded side modules?
If so, would there be a workaround to load a dynamically-uploaded side module larger than 4KB?

**Comment by cuinjune** [comment]
@sbc100 Can I expect this to be fixed at some point?

**Comment by sbc100** [comment]
We would be happy to assist you if you want to propose a change to `loadDynamicLibrary`, but it could be that quickest way to fix is to dig into the problem yourself.

It could be that re-structuring your program to not depend on dynamic loading might be the fastest way to get stuff working.  Can you describe your use case a little?  Does it really depend on the loading symbols by string name at runtime?   Or can we find a way to work around this requirement?

**Comment by cuinjune** [comment]
We are working on making a visual programming software called Purr Data to run in a browser. 
Natively, the software can dynamically load external libraries built by users using `dlopen()`.

Our web version of the software will have a file manager where users can dynamically upload their projects with their external libraries just like how it works natively.

Currently, I could confirm that it is possible to dynamically load `.wasm` side modules using `dlopen()` but it doesn't work in Chrome if the file size is larger than 4KB because `dlopen()` cannot load files asynchronously.

I could load a large-sized file using `loadDynamicLibrary()` but it only works if the file is preloaded.

I would like to try to fix this in emscripten and propose a change but I would still appreciate it if you can guide me on how I should approach this.

1. It looks like `dlopen()` internally calls `loadDynamicLibrary()`. What if I add `loadAsync: true` flags when calling the function? Would it make `dlopen()` to load large-sized files and dynamically?

2. What makes it possible for `dlopen()` to load files dynamically whereas `loadDynamicLibrary()` can't? Would it be because of the `loadAsync: true` flag which uses `fetchBinary()` instead of `readBinary()`? (former uses `fetch()` and latter uses `nodeFS['readFileSync']()`)

**Comment by sbc100** [comment]
I see.  Your use case does indeed look like one that requires genuine runtime dynamic linking.  You are effectively loading plugins into your program at runtime.

1.  dlopen in inherently synchronous and WebAssembly module instantiation in inherently asynchronous (at least for modules over 4k).  Those two facts are not changeable.  However emscripten does have a feature called asyncify which allows can transform seeming synchronous native calls into asynchronous ones: https://emscripten.org/docs/porting/asyncify.html.  So it should be possible in theory for dlopen to use `loadAsync: true` when running in asyncify mode.

2. I have clues here I'm afraid.   This is certainly something that seems fixable.

3. I'm not familiar with that code I'm afraid.

One general piece of advice is perhaps to look at your code in two parts: (1) that native code.  (2) the web/JS part that run the native code.   It might make more sense on the web do to all your plugin loading on the JS side where you can make async calls and use loadDynamicLibrary.   It might also be possible then to avoid the filesystem completely and just load code from URLs.   The filesystem is there for native emulation of stuff like fopen/fread etc.  But if you just want to download and load code in JS then perhaps you don't need to both to write your code into the filesytem at all.

**Comment by cuinjune** [comment]
@sbc100 Thank you so much for your reply. It was super helpful!

I finally got it working! I needed to also pass `fs: FS` flag into the`loadDynamicLibrary()` function.
Here's the full working example code:

**main.c:**
```
#include <stdio.h>
#include <emscripten.h>

EM_JS(void, doLoadLibrary, (), {
    Asyncify.handleAsync(async() => {
    try {
        await loadDynamicLibrary('side.wasm', { loadAsync: true, global: true, nodelete: true, fs: FS });
        console.log('side module size: ' + Module['_side']());
    }
    catch (error) {
        console.log(error);
    }
});
    });

EMSCRIPTEN_KEEPALIVE
void loadLibrary() {
    printf("before\n");
    doLoadLibrary();
    printf("after\n");
}

int main()
{
}
```
**side.c:**
```
#define SIZE 5000
char dummy[SIZE] = {};

int side(void)
{
    return SIZE;
}
```
**index.html:**
```
<!DOCTYPE html>
<html>

<head>
</head>

<body>
  <input id="uploadLibrary" type="file" />
  <button id="loadLibrary">loadLibrary</button>
  <script>

    function uploadLibrary() {
      var files = this.files;
      if (files.length === 0) {
        console.log("No file is selected");
        return;
      }
      var file = files[0];
      var reader = new FileReader();
      reader.onload = function () {
        var data = new Uint8Array(reader.result);
        Module["FS_createDataFile"]("/", file.name, data, true, true, true);
      };
      reader.readAsArrayBuffer(file);
    }
    
    function loadLibrary() {
      Module._loadLibrary();
    }

    document.getElementById("uploadLibrary").addEventListener("change", uploadLibrary, false);
    document.getElementById("loadLibrary").addEventListener("click", loadLibrary, false);
  </script>
  <script async src="main.js"></script>
</body>

</html>
```
**Makefile:**
```
all: clean
	mkdir -p side
	emcc side.c -s SIDE_MODULE=1 -o side/side.wasm
	emcc main.c -s MAIN_MODULE=1 -s ASYNCIFY -s "ASYNCIFY_IMPORTS=['doLoadLibrary']" -s FORCE_FILESYSTEM=1 -s "EXTRA_EXPORTED_RUNTIME_METHODS=['FS']" -o main.html
	python3 -m http.server 8080

clean:
	rm -rf side main.html main.js main.wasm main.data
```
In your browser, click "Choose File" and select the `side/side.wasm` file, and then click "loadLibrary" button.
You will see the following result in the console:

```
before
side module size: 5000
after
```

Thank you so much for your help so far. You can close this issue if you want.

**Comment by sbc100** [comment]
Thats great news!    So glad we figured out a method that worked.

Btw how are you then fishing the symbols out of the code?    I think we still probably want to make `RTLD_DEFAULT` work with `dlsym` to enable this.. but that can be different issue perhaps?

**Comment by cuinjune** [comment]
As you can see from my code, I'm currently using `Module['_side']()` to call the function from a loaded side module and this way seems to work fine.

But if I try to call this function using `dlsym` and `RTLD_DEFAULT`, it was not able to call the function.

Here's an example code you can test this. 
(you just need to replace `main.c` from my previous example with the following)

**main.c:**
```
#include <stdio.h>
#include <dlfcn.h>
#include <emscripten.h>

EM_JS(void, doLoadLibrary, (), {
    Asyncify.handleAsync(async() => {
    try {
        await loadDynamicLibrary('side.wasm', { loadAsync: true, global: true, nodelete: true, fs: FS });
    }
    catch (error) {
        console.log(error);
    }
});
    });

EMSCRIPTEN_KEEPALIVE
void loadLibrary() {
    printf("before\n");
    doLoadLibrary();
    printf("after\n");
    typedef int (*func_t)(void);
    func_t func = (func_t)dlsym(RTLD_DEFAULT, "side");
    if (!func) {
        printf("failed to find the method\n");
        return;
    }
    printf("side module size: %d byte\n", func());
}

int main()
{
}
```
**result:**
```
before
after
failed to find the method

```
So I guess this needs to be fixed in the future although it is currently still possible to call the function with `Module['_side']()`.

**Comment by goldwaving** [comment]
I've been struggling with this as well trying to add the LAME MP3 encoder to GoldWave Infinity.  We need to keep LAME separate to ensure we comply with the LGPL.

The documentation (at https://github.com/emscripten-core/emscripten/wiki/Linking) gives the impression that dynamic linking is straight forward.  For real world apps, it is more complicated.  To help others in the future, I'd recommend updating the documentation to:

- give more emphasis on complications related to +4K modules and steps to solve them
- mention and explain why dlopen cannot be used in a thread to overcome that problem
- mention the drawbacks of preloading (not ideal when there are many large modules/plug-ins that may not be needed until later or at all)
- mention and document loadDynamicLibrary (with "fs: FS" for preloaded), which seems to be the only way to truly load larger modules dynamically.
- use RTLD_DEFAULT in dlsym and add leading underscore as required
- or for non-global, use handle returned by loadDynamicLibrary in dlsym, but no leading underscore

**Comment by kripken** [comment]
A PR with doc improvements would be welcome @goldwaving . Or if you do not have time, opening a new issue with your comments might be better, so that others can see it.

**Comment by mewalig** [comment]
This is insanely frustrating because we ran an initial test using `dlopen()` and it just so happened that our test .so file was smaller than 4k and so it worked perfectly. Have spent more than the entire day trying to reason out what could be going wrong when we drop our "real" .so file in and finally find out it has to do with the size of the .so and that, literally, if I change a literal value in my c code from "this string" to "this string has more chars", dlopen() goes from succeeding to failing which to anyone without peculiar knowledge of emscripten's bowels, just makes no sense at all

It would be better to simply not pretend to support `dlopen()` until it properly works, or works for at least a reasonably sized file, than to have it work but only for use cases that are useless 99% of the time. At the very least, please change the documentation so that it is not misleading-- which it entirely is today-- and to clearly spell out those very particular specific cases where this might actually work. This is a very easy problem to reproduce and to clearly define the limits and boundaries for-- can we not spend the few minutes it takes to add some clarification to the documentation to prevent others from unnecessarily going down this dark rabbit hole?

**Comment by AndyC-NN** [comment]
The easy way to actually do this is to load the module with WebAssembly.instantiate in the javascript side. Use this tutorial to get beyond the 4K limit, https://developers.google.com/web/updates/2018/04/loading-wasm Once the module is loaded, get the function this way:

side = instance.exports._Z4sidev;

So we now have a javascript function of side but this now seems unusable to use in C++, until you realise that Emscripten has a handy function to get the function pointers for C/C++ called addFunction:

sidePtr = addFunction(factorial, 'vi');

The 'vi' to addFunction is really not that important here. It can be anything. That passes back an int that is a function pointer. Some of you will now wonder how can the int be a pointer? Under the hood of Emscripten compiling all functions in C/C++ after the code is compiled in WASM will be an int. With that knowledge you can just do this in C/C++:

typedef int (*side_function)();
side_function side = (side_function)sidePtr;
int size = side();

And somehow compiling with Emscripten this all works great!

**Comment by aisnote** [comment]
@cuinjune  did you meet this kind of issue before?

Seems it is caused by **my.wasm** building settings?
------------------------------
![image](https://user-images.githubusercontent.com/1201634/188051750-02a05fa4-0011-4f0d-b72d-4b0ce2acb624.png)
------------------------------

```c
EM_JS(void, doLoadLibrary, (), {
    Asyncify.handleAsync(async() => {
        try
        {
            await loadDynamicLibrary('ff/lib/my.wasm', {loadAsync : true, global : true, nodelete : true, fs : FS});
            //console.log('side module size: ' + Module['_side']());
        }
        catch (error)
        {
            console.log(error);
        }
    });
});
```

**Comment by sbc100** [comment]
@pd-l2ork, it seems likely that the change that broke your solution was #19310.    This change moved the filesystem reading part of library loading into native code which is now part of `dlopen`.    The behaviour of `loadDynamicLibrary` was also changed, on the assumption that this was an internal function with the public interface being `dlopen`.
My advice would be to switch to using `dlopen` directly, unless there is some fundamental reason why that doesn't work for for you?

By the way the lines in your binary such as `{"filename":"/pd-l2ork/extra/maxlib/average.wasm","start":15274628,"end":15276992}` are the result of the `--preload-file` flag. 

A few other (unrelated) notes on your command line flags.
- You might want to consider switching from `--preload-file` to `--embed-file` which can be more efficient and avoids a copy.
- The `--no-heap-copy` flag is not longer needed.   Its been the default for a long time now.
- The EXPORTED_RUNTIME_METHODS and ASYNCIFY_IMPORTS flags can just take simple comma separated lists without any brackets or quoting, which can make you command line simpler.  e.g. simply `-sEXPORTED_RUNTIME_METHODS=FS`

**Comment by pd-l2ork** [comment]
Thank you @sbc100 for your reply. I think the original issue is that many of these objects are larger than 4KB (see the thread title). Has this been resolved? If so, would you be able to provide a skeleton example function that does the same replacing the __PD_loadlib one? Thank you again for your help.

**Comment by sbc100** [comment]
Chrome has never supported loading module >4kb synchronously.    The only way I know of that works around that would be via preloading of the modules, which `--preload-file` and `--embed-file` should both take care of.   

See `preloadedWasm` in `library_dylink.js` for how this work.

**Comment by pd-l2ork** [comment]
Thanks again @sbc100 . So, now I tried changing code to use dlopen (I tried transplanting the original code by @cuinjune ) and now the system is unable to locate the file. Is dlopen trying to search --preload-file database/list (or whatever it is), or is it trying to locate the file on the local drive where the server is running? If it is going through the list, I have objects that share the same name (but are in different subfolders). How could one distinguish between the two, when the code above shows only referencing the actual filename.wasm? In other words, if, for instance the stored object is:

```
{"filename":"/pd-l2ork/extra/maxlib/average.wasm","start":15274628,"end":15276992}
```

 should the file to be loaded be referenced as:

- average.wasm
- pd-l2ork/extra/maxlib/average.wasm or
- /pd-l2ork/extra/maxlib/average.wasm (note the starting slash)

And what is this location in reference to? To the folder where the server was originally started?

**Comment by pd-l2ork** [comment]
In follow-up to my previous question, how would dlopen example code look inside js? When I try to compile js code with dlopen function, it says it does not recognize the function. Here's the original JS code using loadDynamicLibrary:

```
EM_JS(int, __Pd_loadLib, (const char *filename, const char *symname), {
    return Asyncify.handleAsync(async () => {
        try {
            await loadDynamicLibrary(UTF8ToString(filename), {loadAsync: true, global: true, nodelete: true, fs: FS});
            var makeout = Module['_' + UTF8ToString(symname)];
            if (typeof makeout === "function") {
                makeout();
                console.log("...success");
                return 1; // success
            }
            else {
                console.log("...no function found");
                return -1; // couldn't find the function
            }
        }
        catch (error) {
            console.log("...failed");
            console.error(error);
            return 0; // couldn't load the external
        }
    });
});
```

The C version of dlopen is seen in the top post in this thread:

```
    void *handle;
    typedef int (*func_t)(int);

    handle = dlopen("side.wasm", RTLD_NOW);

    if (!handle) {
        printf("failed to open the library\n");
        return 0;
    }
    func_t func = (func_t)dlsym(handle, "side");

    if (!func) {
        printf("failed to find the method\n");
        dlclose(handle);
        return 0;
    }
    printf("side module size: %d byte\n", func(1));
```

Note this is a part of the main, so I am not even sure how this would look as a separate C function, since I am quite sure the handle would disappear at the end of that function call.

So, how would this dlopen code look inside a js call, like the one above that is wrapped into EM_JS?

**Comment by sbc100** [comment]
I believe with `pd-l2ork/extra/maxlib/average.wasm` or `/pd-l2ork/extra/maxlib/average.wasm` should work.  In the past did `average.wasm` previously work?  I would be surprised if that was the case.

dlopen is a native call so if you don't need to use an EM_JS function at all to call it.   You can, for example, just write a function that calls dlopen + dlsym and then returns that function pointer (i.e. the result of dlsym).

**Comment by kleisauke** [comment]
FWIW, Chrome extended this limit to 8MB in version 115, see commit https://github.com/chromium/chromium/commit/d1a1a8fdd9b28feaf5d3accc0578092444a1083e.

**Comment by pd-l2ork** [comment]
@sbc100 @kripken OK, so I was able to verify that the new object is indeed loading the setup function. However, once the setup is loaded, the next is instantiation (average_new function located inside a separate .c file dedicated to this object/plugin that is to be dynamically loaded). What I found out is that if this function is declared as static, I end up having the Uncaught RuntimeError: null function or function signature mismatch error. If I recompile the object with that function not being static, everything works. This is a problem, because there are over a thousand of 3rd party objects, some that may share functions that are named the same, so it would be a really bad idea to keep all those functions in a global space. More so, the same source is used to compiile a native version o the software, which makes an additional argument against removing static declaration of those variables. So, how does one ensure that you can dlopen external object/library/plugin that may have its own static variables that should be exposed within the browser?

**Comment by sbc100** [comment]
As far as I know there is no way to access at `static` C function from the DLL.   `static` functions are internal to the object file and are not part of the DLL/shared library symbol table.   I believe this is true on standard desktop systems such as linux, macOS and windows too.   Are you saying you are able to use dlopen/dlsym to load `static` functions in some other platform?

**Comment by pd-l2ork** [comment]
This code is actively maintained on desktop and runs on Windows, Linux, and OSX. The average_new function being declared static is no problem because (I think--I am not the original author of this architecture) of the following:

1) average_setup function builds a list of methods object has, and stores them in a class declaration (this is c code, so class is in reality a structure), that can be then referenced later if we instantiate the same object later

2) its instantiation function is invoked through that method, which is a pointer to the static function

As I mentioned, this works fine on all desktop platforms, so I am unsure why it is not working here.

The same code has also worked just fine on the previous version of emscripten that relied upon loadDynamicLibrary.

**Comment by sbc100** [comment]
So IIUC there is some kind of setup function that returns at struct containing function pointers?  That should work fine.  Are you able to run the setup function?  What does it return?   Perhaps if you should share an example of the exact code that is failing that might help (in as simple a form as possible, if you can).

**Comment by pd-l2ork** [comment]
The setup function runs fine. I know this because when run it prints out a message. The average_new is never reached (I have another printout message on top of it). I wish I could simplify this to make it more easy to share. Alas, this may take me a couple of days just to come up with such an abridged version. The overall code base of this thing is humongous--git suggests over 1.5M lines of code, although some of this is definitely due to a collection of images.

**Comment by sbc100** [comment]
How is `average_new` supposed to be called?  Is it called via `dlsym`?  If not, then how/when is it called?   Is it just a the single setup function looked up via `dlsym` or are there other?

**Comment by pd-l2ork** [comment]
Here's a general breakdown:

The program (pd-l2ork) reads a text file (a.k.a. patch) that contains visual dataflow code split into visual object creation (think of it as boxes with names, each with particular function), object connection, and encapsulation. Here, we are only concerned with object creation.

Note that everything I am about to share works perfectly in desktop environments and has worked fine (there were still bugs to work out) in emscripten release sometime in 2021.

Below are key structures and functions:

Core:
- class_new: creates a new class (which is a structure) that contains a list of pointers to various functions inside any object, built-in or dynamically loaded):
```
t_class *class_new(t_symbol *s, t_newmethod newmethod, t_method freemethod,
    size_t size, int flags, t_atomtype type1, ...)
{
    va_list ap;
    t_atomtype vec[MAXPDARG+1], *vp = vec;
    int count = 0;
    t_class *c;
    int typeflag = flags & CLASS_TYPEMASK;
    if (!typeflag) typeflag = CLASS_PATCHABLE;
    *vp = type1;

    va_start(ap, type1);
    while (*vp)
    {
        if (count == MAXPDARG)
        {
            error("class %s: sorry: only %d args typechecked; use A_GIMME",
                s->s_name, MAXPDARG);
            break;
        }
        vp++;
        count++;
        *vp = va_arg(ap, t_atomtype);
    }
    va_end(ap);
    if (pd_objectmaker && newmethod)
    {
        post("class_new newmethod %lx", newmethod);
            /* add a "new" method by the name specified by the object */
   // this is a creation method assigned to this class based on obtained data
  // pd_objectmaker is t_pd struct in charge of creating new objects and to
  // whom creation requests are sent.
        class_addmethod(pd_objectmaker, (t_method)newmethod, s,
            vec[0], vec[1], vec[2], vec[3], vec[4], vec[5]);
        if (class_loadsym)
        {
                /* if we're loading an extern it might have been invoked by a
                longer file name; in this case, make this an admissible name
                too. */
            char *loadstring = class_loadsym->s_name;
            int l1 = strlen(s->s_name), l2 = strlen(loadstring);
            if (l2 > l1 && !strcmp(s->s_name, loadstring + (l2 - l1)))
                class_addmethod(pd_objectmaker, (t_method)newmethod,
                    class_loadsym,
                    vec[0], vec[1], vec[2], vec[3], vec[4], vec[5]);
        }
    }
    c = (t_class *)t_getbytes(sizeof(*c));
    c->c_name = c->c_helpname = s;
    c->c_size = size;
    c->c_methods = t_getbytes(0);
    c->c_nmethod = 0;
    c->c_freemethod = (t_method)freemethod;
    c->c_bangmethod = pd_defaultbang;
    c->c_pointermethod = pd_defaultpointer;
    c->c_floatmethod = pd_defaultfloat;
    c->c_symbolmethod = pd_defaultsymbol;
    c->c_blobmethod = pd_defaultblob; /* MP 20061226 blob type */
    c->c_listmethod = pd_defaultlist;
    c->c_anymethod = pd_defaultanything;
    c->c_wb = (typeflag == CLASS_PATCHABLE ? &text_widgetbehavior : 0);
    c->c_pwb = 0;
    c->c_firstin = ((flags & CLASS_NOINLET) == 0);
    c->c_patchable = (typeflag == CLASS_PATCHABLE);
    c->c_gobj = (typeflag >= CLASS_GOBJ);
    c->c_drawcommand = 0;
    c->c_floatsignalin = 0;
    c->c_externdir = class_extern_dir;
    c->c_savefn = (typeflag == CLASS_PATCHABLE ? text_save : class_nosavefn);

    classtable_register(c);
    post("class:%s newmethod=%lx", c->c_name->s_name, newmethod);
    return (c);
}
```

Core or external objects:
- object_setup function (each object has its own name, e.g. average_setup, or something_setup) that is called once the object is found (either  built-in or dlopen-ed, the desktop version also uses dlopen). Below is an example of the average_setup:

```
void average_setup(void)
{
   // this creates the class, passing it creation function, freeing function, and optional arguments
    average_class = class_new(gensym("average"), (t_newmethod)average_new,
    	(t_method)average_free, sizeof(t_average), 0, A_DEFFLOAT, 0);
   // I added this to post the function address
    post("average_setup %lx", (t_method)average_new);
  // this adds additional methods, all of which are statically defined
    class_addmethod(average_class, (t_method)average_reset, gensym("reset"), 0);
    class_addmethod(average_class, (t_method)average_linear, gensym("linear"), 0);
    class_addmethod(average_class, (t_method)average_geometric, gensym("geometric"), 0);
    class_addmethod(average_class, (t_method)average_weight, gensym("weight"), 0);
    class_addfloat(average_class, average_float);
	class_addmethod(average_class, (t_method)average_index, gensym("index"), A_FLOAT, 0);

   // version is a const char text that is visible when I run my code, so the code gets this far
    logpost(NULL, 4, version);
}
```
- Here's the average_new function that, like all other objects, returns pointer to its instance
```
static void *average_new(t_floatarg f)
{
   // this first post that should print to console is never reached
	post("average_new");
	t_average *x = (t_average *)pd_new(average_class);
	x->x_inindex = inlet_new(&x->x_ob, &x->x_ob.ob_pd, gensym("float"), gensym("index"));
	x->x_outfloat = outlet_new(&x->x_ob, gensym("float"));
	x->x_outtendency = outlet_new(&x->x_ob, gensym("float"));

		/* zero out the array */
	x->x_index = (t_int)f;
	if (x->x_index < 1)
		x->x_index = 1;
	x->x_input = (t_float *)calloc(x->x_index,sizeof(t_float));

	x->x_inpointer = 0;
	x->x_average = 0;
	x->x_mode = 0;
    return (void *)x;
}
```

- function responsible for loading the function and then running the setup:

```
static int sys_do_load_lib(t_canvas *canvas, const char *objectname,
    const char *path)
{
 // cutting bunch of irrelevant stuff
 // ...
// this is what it used to be and what used to call EM_JS __Pd_loadLib call that relied on the loadDynamicLibrary
// now it is commented out
/*
#ifdef __EMSCRIPTEN__
    post("__Pd_loadLib <%s> <%s>", filename, symname);
    int res = __Pd_loadLib(filename, symname);
    if (ret == 1) {
        post("...success");
        (*makeout)();
        class_set_extern_dir(&s_);
        return (1);
    }
    else if (ret == 0) {
        verbose(1, "%s: couldn't load", filename);
        post("...couldn't load");
        class_set_extern_dir(&s_);
        return (0);
    }
    else { //ret is -1, object was opened but no setup function was found
        post("...no setup function found");
        return(0);
    }
#elif _WIN32
*/
#ifdef _WIN32
   // this part is cut, as it does not apply
#elif defined(HAVE_LIBDL) || defined(__FreeBSD__) || defined(__EMSCRIPTEN__) // this is where I've now added EMSCRIPTEN part
   // this is what we use currently
    post("sys_do_load_lib <%s> <%s>", filename, symname);
    dlobj = dlopen(filename, RTLD_NOW | RTLD_GLOBAL);
    if (!dlobj)
    {
        error("%s: %s", filename, dlerror());
        class_set_extern_dir(&s_);
        return (0);
    }
    makeout = (t_xxx)dlsym(dlobj,  symname);
    if(!makeout)
        makeout = (t_xxx)dlsym(dlobj,  "setup");
#else
#warning "No dynamic loading mechanism specified, \
    libdl or WIN32 required for loading externals!"
#endif

    if (!makeout)
    {
        error("load_object: Symbol \"%s\" not found", symname);
        class_set_extern_dir(&s_);
        return 0;
    }
    (*makeout)();
    class_set_extern_dir(&s_);
    post("...sys_do_load_lib success");
    return (1);
}
```
When creating objects that come built-in with the core pd-l2ork package (unlike 3rd party objects), everything is created just fine. However, when trying to create dynamically loaded object, this is something that used to work fine around 2021 and now does not, with only notable change between now and then being emscripten version update. So, I set out to adapt loading using dlopen and here's the current status:

1. Core pd-l2ork reads the object creation request from the patch file. It searches the existing class list and if it does not find it, tries to look for an external. This eventually triggers sys_do_load_lib, which in turn triggers average_setup, which triggers class_new, and then invokes creation of the actual object.

The resulting relevant console printout excerpt shows what worked and where it fails:

```
// this is the call when pd-l2ork does not know what to instantiate, so it calls this to trigger search (1 stands for a
// number of arguments, in this case it is 10. this means that the object will be giving a running average over
// the last 10 numbers
main.js:298 new_anything <maxlib/average> 1
// here you see the call with proper location and setup function
main.js:298 sys_do_load_lib <pd-l2ork-web/extra/maxlib/average.wasm> <average_setup>
// here we see the address of the new method which is 2674
main.js:298 class_new newmethod 2674
// here we see methods being added (we have both the long and short version here, so both versions work)
main.js:298 addmethod total=297 <average>
main.js:298 addmethod total=298 <maxlib/average>
// here we can see that class_new newmethod has indeed correct address matching the one above
main.js:298 class:average newmethod=2674
// here we see what the address of the newmethod is coming from the average_setup
// this also proves that average_setup is found and is triggered
main.js:298 average_setup 2674
// here you see how other methods (static functions inside average.c file) are being added, so that they can be referenced, so this tells us class has indeed gotten new methods
main.js:298 addmethod total=1 <reset>
main.js:298 addmethod total=2 <linear>
main.js:298 addmethod total=3 <geometric>
main.js:298 addmethod total=4 <weight>
main.js:298 addmethod total=5 <index>
// this one is also called from average_setup via logpost (level 4 in terms of verbosity)
main.js:298 verbose(4): average v0.2, written by Olaf Matthes <olaf.matthes@gmx.de>, revised by Ivica Ico Bukvic <ico@vt.edu>
// this tells us that sys_do_load_lib has successfully completed
main.js:298 ...sys_do_load_lib success
// now that the class has been properly created, we call new_anything again
// this time it should be able to create an object because it has its class in its class list
main.js:298 ...new_anything typedmess <maxlib/average> 1
// unfortunately, this is what happens if average_new is a static function (typedmess is simply resending the patch text asking to instantiate the object)
// if I make average not static, the object instantiates fine but that is not an option,
// as doing so will very likely create address clashes among over a thousand of 3rd party objects
main.wasm:0x1359d1 Uncaught (in promise) RuntimeError: null function or function signature mismatch
    at main.wasm:0x1359d1
    at main.wasm:0x135344
    at main.wasm:0x13581d
    at main.wasm:0x130275
    at main.wasm:0x12169d
    at main.wasm:0x12155d
    at main.wasm:0x13557b
    at main.wasm:0x130275
    at main.wasm:0x133a73
    at main.wasm:0x133b0e
$pd_typedmess @ main.wasm:0x1359d1
$new_anything @ main.wasm:0x135344
$pd_typedmess @ main.wasm:0x13581d
$binbuf_eval @ main.wasm:0x130275
$func1654 @ main.wasm:0x12169d
$canvas_obj @ main.wasm:0x12155d
$pd_typedmess @ main.wasm:0x13557b
$binbuf_eval @ main.wasm:0x130275
$binbuf_evalfile @ main.wasm:0x133a73
$glob_evalfile @ main.wasm:0x133b0e
$libpd_openfile @ main.wasm:0x187a73
$_ZN2Pd9openPatchERKNSt3__212basic_stringIcNS0_11char_traitsIcEENS0_9allocatorIcEEEES8_ @ main.wasm:0xaa48b
$_ZN10emscripten8internal13MethodInvokerIM2PdFvRKNSt3__212basic_stringIcNS3_11char_traitsIcEENS3_9allocatorIcEEEESB_EvPS2_JSB_SB_EE6invokeERKSD_SE_PNS0_11BindingTypeIS9_vEUt_ESL_ @ main.wasm:0xa920a
Pd$openPatch @ VM121:10
openPatch @ main.js:2094
init @ main.js:2153
await in init (async)
mainInit @ main.js:840
__mainInit @ main.js:1
$__original_main @ main.wasm:0xa8817
$main @ main.wasm:0xa9488
callMain @ main.js:1
doRun @ main.js:1
run @ main.js:1
runCaller @ main.js:1
removeRunDependency @ main.js:1
processPackageData @ main.js:1
(anonymous) @ main.js:1
xhr.onload @ main.js:1
load (async)
fetchRemotePackage @ main.js:1
loadPackage @ main.js:1
(anonymous) @ main.js:1
(anonymous) @ main.js:1
```

Below is the relevant build output:

average.c:
```
emcc -DPD -I/home/l2orkist/Downloads/pd-l2ork/emscripten/../pd/src -Wall -W -Wno-unused-parameter -s SIDE_MODULE=1 -O3 -o "average.o" -c "average.c"
average.c:218:22: warning: format string is not a string literal (potentially insecure) [-Wformat-security]
  218 |     logpost(NULL, 4, version);
      |                      ^~~~~~~
average.c:218:22: note: treat the string as an argument to avoid this
  218 |     logpost(NULL, 4, version);
      |                      ^
      |                      "%s", 
1 warning generated.
emcc -s SIDE_MODULE=1 -O3 -o "average.wasm" "average.o"   
chmod a-x "average.wasm"
```

main (combination of core and all bundled externals):
```
emcc -I/home/l2orkist/Downloads/pd-l2ork/emscripten/build/../../pd/src -I/home/l2orkist/Downloads/pd-l2ork/emscripten/build/../../libpd/libpd_wrapper -O3 -s MAIN_MODULE=1 --bind -o main.html /home/l2orkist/Downloads/pd-l2ork/emscripten/build/../../emscripten/src/main.cpp /home/l2orkist/Downloads/pd-l2ork/emscripten/build/../../emscripten/src/pd.cpp /home/l2orkist/Downloads/pd-l2ork/emscripten/build/../../emscripten/src/js.cpp \
-s USE_SDL=2 -s ERROR_ON_UNDEFINED_SYMBOLS=0 -s ALLOW_MEMORY_GROWTH=1 \
-s FORCE_FILESYSTEM=1 -s EXPORTED_RUNTIME_METHODS=FS \
--preload-file pd-l2ork-web -L/home/l2orkist/Downloads/pd-l2ork/emscripten/build/../../libpd/libs -lpd -lm
```

**Comment by sbc100** [comment]
Regarding the old EM_JS code, I think I do see a problem there.   The code is expecting symbols from the loaded library to be presend on the global Module object with the underscore prefix.   i.e. `var makeout = Module['_' + UTF8ToString(symname)];`   

These days we use `wasmImports` as the global symbol table.    So that line should be replaced with `var makeout = wasmImports[UTF8ToString(symname)];`.   That is, assuming that `loadDynamicLibrary` succeeds.

However, the dlopen approach should also work and be simpler and more robust over time.   Are the dlsym calls succeeding?   Presumable one of the `gensym` calls is failing?  What does `gensym` do exactly?  Does it call dlsym?

**Comment by sbc100** [comment]
If you want to get  better stack traces can you try building with `-g` (or just `--profiling-funcs`).  That should give function names rather than just:

```
signature mismatch
    at main.wasm:0x1359d1
    at main.wasm:0x135344
    at main.wasm:0x13581d
    at main.wasm:0x130275
    at main.wasm:0x12169d
    at main.wasm:0x12155d
    at main.wasm:0x13557b
    at main.wasm:0x130275
    at main.wasm:0x133a73
    at main.wasm:0x133b0e
```

**Comment by pd-l2ork** [comment]
> These days we use `wasmImports` as the global symbol table. So that line should be replaced with `var makeout = wasmImports[UTF8ToString(symname)];`. That is, assuming that `loadDynamicLibrary` succeeds.

Thank you. I will try to revert back to that implementation and see how it goes.

> However, the dlopen approach should also work and be simpler and more robust over time. Are the dlsym calls succeeding? Presumable one of the `gensym` calls is failing? What does `gensym` do exactly? Does it call dlsym?

gensym is simply generating a symbol (string) that is kept in a string table. This is needed for a number of reasons, including wirelessly connecting objects via internal send and receive objects (among other things).

BTW, dlopen does work up through the setup function. Where it fails is when trying to instantiate average_new, which should've been stored in class_new, and whose address, per my debug printouts, appears to be tracking across multiple function calls.

**Comment by pd-l2ork** [comment]
> If you want to get better stack traces can you try building with `-g` (or just `--profiling-funcs`). That should give function names rather than just:
> 
> ```
> signature mismatch
>     at main.wasm:0x1359d1
>     at main.wasm:0x135344
>     at main.wasm:0x13581d
>     at main.wasm:0x130275
>     at main.wasm:0x12169d
>     at main.wasm:0x12155d
>     at main.wasm:0x13557b
>     at main.wasm:0x130275
>     at main.wasm:0x133a73
>     at main.wasm:0x133b0e
> ```

I will also make sure to compile with a debug flag and will let you know. Thank you.

**Comment by pd-l2ork** [comment]
So, I am completely baffled. I just went to enable the debug options and I don't think I even did it on all components that comprise the final main wasm file and upon recompiling dlopen worked as expected. I then, went on to disable debugging to revert everything back and was unable to break it again. Eventually, I was able to enable debugging for everything and that ended-up spawning some odd errors, but this is a separate issue at this point. I know for a fact that I cleaned the entire tree before. Could it be that there may have been a stale .o file that could've caused this? I am thinking .o file from a different compile time may generate a different signature, resulting in an incompatible build perhaps? I am at a loss, and my brain is fried... More tomorrow... Thanks again for all your help.

**Comment by pd-l2ork** [comment]
For whatever rason I am now unable to break it in the way it was broken before, but am now running into the debug-enabled version throwing bunch of errors that are not present in an optimized build. Below is an example of the output. Everything loads fine in the browser, but as son as I click, I get the following stream of errors (they repeat, so below is only a snapshot). None of this is a problem when running -O3 build (this one is -g):

```
main.js:56 Aborted(Assertion failed: exceptfds not supported)
printErr @ main.js:56
abort @ main.js:950
assert @ main.js:631
___syscall__newselect @ main.js:5420
$select @ main.wasm:0x28f3a7
$sys_domicrosleep @ main.wasm:0x15ce2e
$sys_pollgui @ main.wasm:0x15eb39
$libpd_process_float @ main.wasm:0x1aa664
$Pd::audioOut(unsigned char*, int) @ main.wasm:0xb90db
$Pd::forwardAudioOut(void*, unsigned char*, int) @ main.wasm:0xb6c3b
$HandleAudioProcess @ main.wasm:0x209500
dynCall @ main.js:1622
SDL2.audio.scriptProcessorNode.onaudioprocess @ main.js:1287
main.js:969 Uncaught RuntimeError: Aborted(Assertion failed: exceptfds not supported)
    at abort (main.js:969:11)
    at assert (main.js:631:5)
    at ___syscall__newselect (main.js:5420:7)
    at select (main.wasm:0x28f3a7)
    at sys_domicrosleep (main.wasm:0x15ce2e)
    at sys_pollgui (main.wasm:0x15eb39)
    at libpd_process_float (main.wasm:0x1aa664)
    at Pd::audioOut(unsigned char*, int) (main.wasm:0xb90db)
    at Pd::forwardAudioOut(void*, unsigned char*, int) (main.wasm:0xb6c3b)
    at HandleAudioProcess (main.wasm:0x209500)
abort @ main.js:969
assert @ main.js:631
___syscall__newselect @ main.js:5420
$select @ main.wasm:0x28f3a7
$sys_domicrosleep @ main.wasm:0x15ce2e
$sys_pollgui @ main.wasm:0x15eb39
$libpd_process_float @ main.wasm:0x1aa664
$Pd::audioOut(unsigned char*, int) @ main.wasm:0xb90db
$Pd::forwardAudioOut(void*, unsigned char*, int) @ main.wasm:0xb6c3b
$HandleAudioProcess @ main.wasm:0x209500
dynCall @ main.js:1622
SDL2.audio.scriptProcessorNode.onaudioprocess @ main.js:1287
main.js:56 Aborted(Assertion failed: exceptfds not supported)
printErr @ main.js:56
abort @ main.js:950
assert @ main.js:631
___syscall__newselect @ main.js:5420
$select @ main.wasm:0x28f3a7
$sys_domicrosleep @ main.wasm:0x15ce2e
$sys_pollgui @ main.wasm:0x15eb39
$libpd_process_float @ main.wasm:0x1aa664
$Pd::audioOut(unsigned char*, int) @ main.wasm:0xb90db
$Pd::forwardAudioOut(void*, unsigned char*, int) @ main.wasm:0xb6c3b
$HandleAudioProcess @ main.wasm:0x209500
dynCall @ main.js:1622
SDL2.audio.scriptProcessorNode.onaudioprocess @ main.js:1287
main.js:969 Uncaught RuntimeError: Aborted(Assertion failed: exceptfds not supported)
    at abort (main.js:969:11)
    at assert (main.js:631:5)
    at ___syscall__newselect (main.js:5420:7)
    at select (main.wasm:0x28f3a7)
    at sys_domicrosleep (main.wasm:0x15ce2e)
    at sys_pollgui (main.wasm:0x15eb39)
    at libpd_process_float (main.wasm:0x1aa664)
    at Pd::audioOut(unsigned char*, int) (main.wasm:0xb90db)
    at Pd::forwardAudioOut(void*, unsigned char*, int) (main.wasm:0xb6c3b)
    at HandleAudioProcess (main.wasm:0x209500)
abort @ main.js:969
assert @ main.js:631
```

I tried compiling with -s USE_PTHREADS as one of the online resources suggested, but that made things even worse, making the program fail to even load (threw one error that I don't have on hand and stopped right there).

Should I make this a separate issue?

**Comment by pd-l2ork** [comment]
Here is a more on-topic question. So, when I run a patch (a snippet of code) that does not require dlopen, everything works fine. However, if I have a patch that invokes dlopen AND the object in question is supposed to process audio, even though the object now loads correctly, as soon as I do anything on the page with my mouse (e.g. click onto a visual widget, a slider or a button), I get the following error:

```
main.wasm:0x1342f3 Uncaught RuntimeError: memory access out of bounds
    at main.wasm:0x1342f3
    at main.wasm:0x188d52
    at main.wasm:0xaa597
    at main.wasm:0xa9335
    at ClassHandle.Pd$sendFloat [as sendFloat] (eval at newFunc (main.js:1:396680), <anonymous>:10:1)
    at gui_slider_bang (main.js:1407:15)
    at gui_slider_onmousemove (main.js:1440:9)
    at main.js:1615:9
```

The source of error may be different depending on what I pressed, but the end result is always the same. I tried recompiling with ```-s TOTAL_MEMORY=128MB``` instead of ```-s ALLOW_MEMORY_GROWTH=1``` thinking perhaps resizing the heap may screw up addresses which are stored in the class, in case the entire thing is repositioned to a different memory address, but that made no difference.

**Comment by sbc100** [comment]
Looks like a memory corruption issue.  I would recommend a debug build with `-sASSERTIONS` and `--profiling-funcs`.   That might.   You might also want to try `-fsanitize=address`.    It sounds like we are getting off topic for this issue though.

**Comment by sbc100** [comment]
Regarding the original issue:  Chome just increased the limit from 4kb to 8Mb 
 https://github.com/chromium/chromium/commit/d1a1a8fdd9b28feaf5d3accc0578092444a1083e.   So this issue should go away once that hits chrome stable.

**Comment by pd-l2ork** [comment]
So, as a closure to this discussion, I learned two critical things:

1) fixed exceptfds issue by updating the source (see #19951 for more info)
2) learned that -fsanitize=address is the magical thing that solves all my problems. If I don't use this compile flag, I get random out of memory errors that happen all over the code, which is unusual for a fairly mature source base that has been running on other platforms for years. Once I enabled this flag, both -g and -O3 builds run just fine. I imagine perhaps this flag makes address handling stricter, which is essential for the way m_class functions, as described above.

Thanks @sbc100 again for all your time and assistance.

**Comment by Lapjared** [comment]
@cuinjune https://lnkr.fun/90c43e1
